source/Calamari/Kubernetes/Conventions/HelmTemplateValueSourcesCreator.cs

```diff
@@ -0,0 +1,252 @@
+                return Enumerable.Empty<string>();
+
+            var parsedJsonArray = JArray.Parse(templateValueSources);
+
+            var filenames = new List<string>();
+            // we reverse the order of the array so that we maintain the order that sources at the top take higher precendences (i.e. are adding to the --values list later),
+            // however, within a source, the file path order must be maintained (for consistency) so that later file paths take higher precendence 
+            foreach (var (json, index) in parsedJsonArray.Select((json, index) => (json, index)).Reverse())
+            {
+                var tvs = json.ToObject<TemplateValuesSource>();
+                switch (tvs.Type)
+                {
+                    case TemplateValuesSourceType.Chart:
+                        var chartTvs = json.ToObject<ChartTemplateValuesSource>();
+                        var chartFilenames = GenerateAndWriteChartValues(deployment, fileSystem, log, chartTvs.ValuesFilePaths);
+
+                        filenames.AddRange(chartFilenames);
+                        break;
+                    case TemplateValuesSourceType.KeyValues:
+                        var keyValuesTvs = json.ToObject<KeyValuesTemplateValuesSource>();
+                        var keyValueFilename = GenerateAndWriteKeyValues(deployment, fileSystem, keyValuesTvs.Value, index);
+
+                        AddIfNotNull(filenames, keyValueFilename);
+                        break;
+
+                    case TemplateValuesSourceType.Package:
+                        var packageTvs = json.ToObject<PackageTemplateValuesSource>();
+                        var packageFilenames = GenerateAndWritePackageValues(deployment,
+                                                                             fileSystem,
+                                                                             log,
+                                                                             packageTvs.ValuesFilePaths,
+                                                                             packageTvs.PackageId,
+                                                                             packageTvs.PackageName);
+
+                        filenames.AddRange(packageFilenames);
+                        break;
+                    case TemplateValuesSourceType.InlineYaml:
+                        var inlineYamlTvs = json.ToObject<InlineYamlTemplateValuesSource>();
+                        var inlineYamlFilename = GenerateAndWriteInlineYaml(deployment, fileSystem, inlineYamlTvs.Value, index);
+
+                        AddIfNotNull(filenames, inlineYamlFilename);
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
+            }
+
+            return filenames;
+        }
+
+        public static IEnumerable<string> GenerateAndWritePackageValues(RunningDeployment deployment,
```

Comment:
These methods look like they could be pulled into their own classes and registered with autofac keyed by template values source type.  It would make this class (and especially the `ParseTemplateValuesSources` method) a fair bit cleaner and have fewer random public methods here too. 
You might need to make them just accept the `XYZTemplateValuesSource` as a param though.

Reply:
This would also mean that the tests around `HelmTemplateValueSourcesCreator` would be simpler, and could have tests directly for each type's functionality.